src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2616 @@
+                len);
+            SCLogInfo("%02x", buf[2]);
+            goto done;
+        }
+        DNP3ObjHeader *header = (DNP3ObjHeader *)buf;
+        offset += sizeof(DNP3ObjHeader);
+
+        DNP3Object *object = SCCalloc(1, sizeof(*object));
+        if (object != NULL) {
+            object->group = header->group;
+            object->variation = header->variation;
+            TAILQ_INSERT_TAIL(objects, object, next);
+        }
+
+        SCLogDebug("Group: 0x%02x (%d); Variation: 0x%02x (%d); "
+            "Qualifier: 0x%02x (%d)",
+            header->group, header->group,
+            header->variation, header->variation,
+            header->qualifier, header->qualifier);
+        uint8_t prefix_code = DNP3_OBJ_PREFIX(header->qualifier);
+        uint8_t range_code = DNP3_OBJ_RANGE(header->qualifier);
+        object->prefix = DNP3_OBJ_PREFIX(header->qualifier);
+        object->range = DNP3_OBJ_RANGE(header->qualifier);
+
+        int prefix_len = DNP3_OBJ_PREFIX_CODE_MAP[prefix_code];
+        if (prefix_len < 0) {
+            goto unknown;
+        }
+
+        uint32_t count = 0;
+
+        /* IEEE 1815-2012, Table 4-5. */
+        switch (range_code) {
+            case 0x00: {
+                uint8_t start = buf[offset++];
+                uint8_t stop = buf[offset++];
+                count = stop - start + 1;
+                break;
+            }
+            case 0x01: {
+                uint16_t start = dnp3_swap16((uint16_t)(*buf));
+                buf = buf + sizeof(uint16_t);
+                uint16_t stop = dnp3_swap16((uint16_t)(*buf));
+                count = stop - start + 1;
+                break;
+            }
+            case 0x02: {
+                uint32_t start = dnp3_swap32((uint32_t)(*buf));
+                buf = buf + sizeof(uint32_t);
+                uint32_t stop = dnp3_swap32((uint32_t)(*buf));
+                count = stop - start + 1;
+                break;
+            }
+            case 0x06:
+                /* No range field. */
+                count = 0;
+                break;
+            case 0x07:
+                /* 1 octet count of objects. */
+                count = buf[offset++];
+                break;
+            case 0x08: {
+                /* 2 octet count of objects. */
+                count = *(uint16_t *)(buf + offset);
+                offset += sizeof(uint16_t);
+                break;
+            }
+            case 0x03:
+            case 0x04:
+            case 0x05:
+            case 0x09:
+            case 0x0b:
+                SCLogInfo("Range code 0x%02x not yet implemented.", range_code);
+                goto done;
+            default:
+                SCLogInfo("Range code 0x%02x is reserved.", range_code);
+                goto done;
+        }
+
+        if (count == 0) {
+            goto next;
+        }
+
+        /* Get the size of the object - in bits. */
+        int object_size = GROUP_VAR_SIZE_MAP[header->group][header->variation];
```

Comment:
group and variation are both uint8_t right? Since GROUP_VAR_SIZE_MAP is [256][256], it should be safe